src/awkward/operations/structure.py

```diff
@@ -1234,6 +1249,9 @@ def getfunction(layout):
             return None
 
     out = ak._util.recursively_apply(layout, getfunction, pass_depth=False)
+    if dtype is not None:
+        out = strings_astype(out, dtype)
+        out = values_astype(out, dtype)
```

Comment:
Why is this doing both? This would chain the output through `strings_astype`, followed by `values_astype`. Is that really desirable? If `dtype=np.int64`, is `strings_astype` a no-op?

Reply:
If there is a string in the input array then  `out = strings_astype(out, dtype)` converts it (though for booleans to work properly the string needs to be '0' or '1', hence that thing on line 1068).
Where as `out = values_astype(out, dtype)` converts numeric things. 

So `strings_astype` is not a no-op, for example;
with line 1253 
`ak.ones_like([6, 'bird'], dtype=bool)` gives `<Array [True, True] type='2 * union[bool, bool]'>`
without line 1253 we get
`ak.ones_like([6, 'bird'], dtype=bool)` gives `<Array [True, '1'] type='2 * union[bool, string]'>`
I would have thought that either it should not touch the string at all or convert it fully, but I may well have misunderstood?

It could be compacted as;
`out = strings_astype(values_astype(out, dtype), dtype)`

Reply:
That makes sense! I had forgotten that the first word ("strings" or "values") is what it's converting _from_. So this chain could be understood as "convert everything into the specified dtype."

Okay, I'm fine with it. The PR was easy to review, the tests passed, so if you're done, I'll merge it.

Thanks again!

Reply:
Yes, all good to go.